controller/volume_controller.go

```diff
@@ -1029,11 +1029,17 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 		// make sure the volume is detached before automatically salvage replicas
 		if autoSalvage && v.Status.State == types.VolumeStateDetached && !v.Status.IsStandby && !v.Status.RestoreRequired {
 			// There is no need to auto salvage (and reattach) a volume on an unavailable node
+			// If we return error we will loop forever, instead we can do auto salvage even if the volume should be detached
+			// since all auto salvage does is unset the failure state for the replicas
+			// on a single replica volume, if the replica is on a failed node, it will be skipped below
+			// and the volume will remain in the faulted state but as soon as the node of the replica comes back
+			// we can do the auto salvage operation and set the Robustness to Unknown.
+			shouldBeAttached := v.Spec.NodeID != ""
 			isNodeDownOrDeleted, err := vc.ds.IsNodeDownOrDeleted(v.Spec.NodeID)
-			if err != nil {
+			if err != nil && shouldBeAttached {
 				return err
 			}
-			if !isNodeDownOrDeleted {
+			if !isNodeDownOrDeleted || !shouldBeAttached {
```

Comment:
Can we decouple the auto salvage from the node down case or auto reattachment feature? If there is a replica candidate on a running node, we will unset `spec.failedAt`. If this replica is actually no longer reusable, we can fail the volume again next time the volume is attached.

Reply:
I am going for a surgical change for the hotfix, since it's currently hard to reason about the volume controller behavior :)
In this case less change is better, till we have a volume controller that can be better reasoned about :)

Reply:
Got it. Let's re-visit this during the refactor